### DISCUSSION TITLE
What grammar is used for P?

### DISCUSSION THREAD

**Discussion by greggirwin** [p-org/P#634] [Q&A] [answered]
I see sections of grammar specs in the docs, but it's not a grammar I'm familiar with. I might be able to render syntax diagrams for P, but it will be easier if it's one of `[ABNF EBNF McKeeman DBNF WBNF]` as my tool handles those. Otherwise I'd need to translate it manually.

Side note: 20+ years ago I wrote a little FSM dialect, and finding P makes me want to dust it off. :^)

**Comment by ankushdesai** [comment, marked as answer]
Thanks! The grammar in the docs is an adopted version of the ANTLR grammar file here: https://github.com/p-org/P/blob/master/Src/PCompiler/CompilerCore/Parser/PParser.g4

Any help to improve the documentation would be great!

**Reply by greggirwin** [reply]
Any thoughts on the value of the diagram @ankushdesai? It can also be rendered into separate images by rule, for targeted doc inclusion.

**Comment by greggirwin** [comment]
Thanks! I forgot to check g4. We have an experimental metagrammar, but I'll have to dust things off to test.

**Reply by greggirwin** [reply]
Don't remember how to trunc the image in chat here. :^\

**Reply by greggirwin** [reply]
Can delete it if you want.

**Reply by greggirwin** [reply]
Ah, need to include the lexer rules from the separate file.

**Reply by greggirwin** [reply]
Lexer rules included. I just hadn't downloaded it. Looks much nicer now. Zipping to keep chat small. Will delete earlier image.
[PParser.zip](https://github.com/p-org/P/files/12222112/PParser.zip)
